src/aiida_quantumespresso/workflows/xspectra/crystal.py

```diff
@@ -0,0 +1,676 @@
+            scf_params = inputs.scf.pw.parameters.get_dict()
+            ch_inputs = XspectraCoreWorkChain.get_treatment_inputs(treatment=ch_treatment)
+            new_scf_params = recursive_merge(left=scf_params, right=ch_inputs)
+
+            # Set the parameter `xiabs` correctly, now that we know which structure we have
+            new_xs_params = inputs.xs_prod.xspectra.parameters.get_dict()
+            kinds_present = sorted([kind.name for kind in structure.kinds])
+            abs_species_index = kinds_present.index(abs_atom_marker) + 1
+            new_xs_params['INPUT_XSPECTRA']['xiabs'] = abs_species_index
+
+            # Set `starting_magnetization` if we are using an XCH approximation, using the
+            # absorbing species as a reasonable place for the unpaired electron.
+            if ch_treatment == 'xch_smear':
+                new_scf_params['SYSTEM'][f'starting_magnetization({abs_species_index})'] = 1
+
+            core_hole_pseudo = self.inputs.core_hole_pseudos[abs_element]
+            gipaw_pseudo = self.inputs.gipaw_pseudos[abs_element]
+            inputs.scf.pw.pseudos[abs_atom_marker] = core_hole_pseudo
+            inputs.scf.pw.pseudos[abs_element] = gipaw_pseudo
+
+            inputs.scf.pw.parameters = orm.Dict(dict=new_scf_params)
+            inputs.xs_prod.xspectra.parameters = orm.Dict(dict=new_xs_params)
+
+            inputs = prepare_process_inputs(XspectraCoreWorkChain, inputs)
+
+            future = self.submit(XspectraCoreWorkChain, **inputs)
+            self.to_context(xspectra_core_workchains=append_(future))
+            self.report(f'launched XspectraCoreWorkChain for {site}<{future.pk}>')
+
+    def inspect_all_xspectra_core(self):
+        """Check that all the XspectraCoreWorkChain sub-processes finished sucessfully."""
+
+        labels = self.ctx.structures_to_process.keys()
+        work_chain_nodes = self.ctx.xspectra_core_workchains
+        failed_work_chains = []
+        for work_chain, label in zip(work_chain_nodes, labels):
+            if not work_chain.is_finished_ok:
+                failed_work_chains.append(work_chain)
+                self.report(f'XspectraCoreWorkChain for ({label}) failed with exit status {work_chain.exit_status}')
+        if len(failed_work_chains) > 0:
+            return self.exit_codes.ERROR_SUB_PROCESS_FAILED_XSPECTRA
+
+    def results(self):
+        """Compile all output spectra, organise and post-process all computed spectra, and send to outputs."""
+
+        site_labels = self.ctx.structures_to_process.keys()
+        work_chain_nodes = self.ctx.xspectra_core_workchains
+        spectra_nodes = {label : node.outputs.powder_spectrum for label, node in zip(site_labels, work_chain_nodes)}
+        spectra_nodes['metadata'] = {'call_link_label' : 'compile_final_spectra'}
+
+        equivalent_sites_data = self.ctx.equivalent_sites_data
+        elements_list = orm.List(list=self.ctx.elements_list)
```

Comment:
Lose data provenance.

Reply:
I will add something to pass the `elements_list` given in the `WorkChain` inputs to this variable here instead of re-defining it. In order for this to work, we will have to change `elements_list` to be a required input so that it is always defined as a `List` node in the inputs.

Since setting a default value is impossible in this case, we can instead change the `get_builder_from_protocol()` method to define this automatically and let the built-in checks confirm that a list of elements has been provided.